match/tensordata.py

```diff
@@ -402,13 +402,65 @@ def broadcast(self, *shape: int):
 
     def __all_coordinates(self):
         return all_coordinates(self.shape)
+    
+    def sum_along_axes(self, axes: list | int = 0, keepdims: bool = False) -> TensorData:
+        """Return a new TensorData object summed along the axis
+        if coord is (2,3) and sum alon axis=1, then we increment the value at (2,0) in the new tensor by the 
+       value at (2,3) in the original tensor. initialize new tensor to be all zeros with the new shape.
+       keepdims = True by default (as far as algorithm is cocnerned), then keepdims = false will simply call a 
+       reshape
+       for a given coordinate, set all dimensins to 0 for every axis we're trying to sum over
+
+       initialize new tensor with new shape (depending on keepdims) to all 0's
+       for (x1, x2, ...) in orig.all_coordinate()
+       new_coordinate = [0 if dim in axes else val for dim, val in enumerate((x1, x2, ...))]
+       new_tensor[new_coordinate] += orig[(x1, x2, ...)]
+
+       return new_tensor
+       """
+        if isinstance(axes, int):
+            axes = [axes]
+
+        if any(ax < 0 or ax >= len(self.shape) for ax in axes):
+            raise ValueError(f"Axes should be in bounds 0 and {len(self.shape)}-1")
+        
+        new_shape = tuple(1 if dim in axes else val for dim, val in enumerate(self.shape))
+        new_tensor = TensorData(*new_shape, value=0) # Must initialize to 0
+        # for every coordinate in the original tensor
+        for orig_index, orig_coord in enumerate(self.__all_coordinates()):
+            new_coord = tuple(0 if dim in axes else val for dim, val in enumerate(orig_coord))
+            new_coord_index = new_tensor.__multi_to_single_rank_translation(new_coord)
+            new_tensor._data[new_coord_index]._item += self._data[orig_index]._item
+       
+        if not keepdims:
+            # remove the dimensions that were summed out (changed to 1's), i.e., the dimensions in axes
+            new_shape = []
+            for dim, val in enumerate(self.shape):
+                if dim not in axes:
+                    new_shape.append(val)
+            new_tensor.reshape(*new_shape)
+       
+        return new_tensor
 
-    def unbroadcast(self, *shape: int):
-        """Return a new TensorData unbroadcast from current shape to desired shape."""
-        if self.shape == shape:
-            return self
-        # TODO (SAM) : Complete unbroadcast.
-        raise NotImplementedError
+    def unbroadcast(self, *shape: int) -> TensorData:
+        """Return a new TensorData unbroadcast from current shape to desired shape.
+        
+        Reference to this: https://mostafa-samir.github.io/auto-diff-pt2/#unbroadcasting-adjoints
+        """
+        # change to shallow copy
```

Comment:
```suggestion
        # TODO(SAM): change to shallow copy
```

Reply:
It is nice to mark these with "TODO". I have an extension that keeps track of them.

Reply:
Ack.